activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/model/impl/APIProcessInstanceConverter.java

```diff
@@ -34,6 +42,33 @@ public ProcessInstance from(org.activiti.engine.runtime.ProcessInstance internal
         processInstance.setProcessDefinitionKey(internalProcessInstance.getProcessDefinitionKey());
         processInstance.setBusinessKey(internalProcessInstance.getBusinessKey());
         processInstance.setStatus(calculateStatus(internalProcessInstance));
+        processInstance.setProcessDefinitionVersion(internalProcessInstance.getProcessDefinitionVersion());
+        
+        //To do: it is not the best way to search parentProcessId by this method, it will require extra queries!
+        
```

Comment:
It will be a general problem for lazy loaded associations when using converters outside of Activiti engine transaction. This will also create N+1 additional queries to database if we need to resolve child entities for each record. We should be able to solve this by implementing or using a query to join associated entities in one query from the database.

Reply:
Another question: do we really need the `parentProcessInstanceId` or `rootProcessInstanceId` and `callerId`(not sure how to name this one) is enough? If we really need it, maybe we should consider to add an extra property to the internal process instance and store it in the database: similar to what is done for `rootProcessInstanceId`. @igdianov what do you think about that? @salaboy you may be interested in this conversation as well.

Reply:
rootProcessInstanceId - is present but not clear if it is supported (filled) at all.
parentId (parentProcessInstanceId) we will need for sure to get list if subprocesses...
What is callerId and do we need it is a question...

Reply:
@erdemedeiros It makes sense to provide parentProcessInstanceId for query, audit services. This use case demonstrates several problems. Problem 1 is that internal api converters need to run inside Activiti transaction in order to be able to resolve lazily associated entities, i.e. `superExecution`, to get its parent process instance id. Problem 2 is that resolving lazy associations this way for each record will create N+1 additional queries to the engine database from many places (aggregating events, apis, queries) that we cannot control because of Problem 1. 

The parent process instance seems to be a common property shared and used by entity and event classes, so this particular use case could be solved by making changes to the execution entity manager query in Activiti engine to always eagerly load super execution entity via join query if it exists. This should have less performance impact overall.